concepts/bits/about.md

```diff
@@ -0,0 +1,263 @@
+For example
+
+```c
+#include <stdio.h>
+
+int main() {
+    char a = 'a';
+    char b = 'b';
+    char c = 'c';
+    // we will use "a" for the "key"
+    char first = a ^ 'a';
+    char second = b ^ 'a';
+    char third = c ^ 'a';
+    // prints three "dots" for unprintable characters
+    printf("%c %c %c\n", first, second, third);
+    // prints the ascii values 0 3 2
+    printf("%d %d %d\n", first, second, third);
+    first = first ^ 'a';
+    second = second ^ 'a';
+    third = third ^ 'a';
+    // prints a b c
+    printf("%c %c %c\n", first, second, third);
+    // prints the ascii values 97 98 99
+    printf("%d %d %d\n", first, second, third);
+}
+```
+
+### Bitwise shift right operator: a way to halve a value
+
+`>>` is the bitwise operator for shifting right.
+Usually, for every position that a value's bits are shifted right, the result is the value divided by two.
+For example
+
+```c
+#include <stdio.h>
+
+int main() {
+    int number = 16;
+    number >>= 2;
+    // prints 4, or 16 / 2 = 8 / 2 = 4
+    printf("%d\n", number);
+}
+```
+
+If the value of the right operand is negative or is greater than or equal to the width of the promoted left operand, the behavior is undefined.
+That also applies to the left shift operator.
+When a signed integer is used instead of an unsigned integer, then the signedness may also factor in.
+The following example shows behavior on a system when shifting too far to the left or right for a signed integer.
+
+```c
+#include <stdio.h>
+
+int main() {
+    int number = 16;
+    number <<= 250;
+    // prints 1073741824
+    printf("%d\n", number);
+    number <<= 1;
+    // prints -2147483648
+    printf("%d\n", number);
+    number <<= 1;
+    // prints 0
+    printf("%d\n", number);
+    // total shifts are now 256, done incrementally
+    number <<= 4;
+    // prints 0
+    printf("%d\n", number);
+    number = 16;
+    // prints 16 when 256 total shifts done at once
+    printf("%d\n", number << 256);
+    // prints 0 when bits shifted right far enough
+    printf("%d\n", number >> 5);    
+    // still prints 0
+    printf("%d\n", number >> 255);
+    // prints 16 when shifted right far enough to "wrap around"
+    printf("%d\n", number >> 256);
+}
+```
+
+Since the behavior is undefined, a different system may yield different results.
+It is important when shifting bits to understand the confines of the data type being used.
+
+## Bit Fields
+
+We've seen how bits packed into a simple data type can be addressed by their position.
+Bits packed into a `struct` can be addressed by name.
+Such a data type is called a `bit field`.
+A use case for `bit fields` can be when a collection of ON/OFF flags is needed and the minimizing of memory use is desired.
+
+```c
+#include <stdio.h>
+
+int main() {
+
+    struct lights {
+        unsigned int number_of_lights;
+        unsigned int red   : 1;
+        unsigned int green : 1;
+        unsigned int blue  : 1;
+    };
```

Comment:
May be good to have an example, or describe, that the number after the colon can be more than 1

Reply:
That's a more complex permutation that may have a place in a "second edition" of these things. My goal with the first iteration of writing the concepts is to simply explain things based on what I think could be helpful to people who are fairly new to the concept. In other words, my goal is to expand horizontally across the subject matter, as there are a lot of subjects to cover. After that phase is complete, the subjects having more detail can be drilled down more vertically.

For example, on covering endianness... Since the purpose of concepts is to tie in with concept exercises, unless we have a concept exercise that deals with endianness, I'm not seeing it as a priority to cover it. I agree it's nice-to-have info, but I'd rather get the more relevant info out there sooner than later, and then fill in the finer points after that.

Reply:
To put if more succinctly: first breadth, then depth.